### DISCUSSION TITLE
x-reactions 自定义使用请教

### DISCUSSION THREAD

**Discussion by prgrmrwy** [alibaba/formily#1461] [Q&A 2.x] [answered]
这种用法会提示引用错误，但是实际已经传入了
```
{
   'x-reactions': {
      dependencies: ['xxx'],
      fulfill: {
          run: '{{useAsyncDataSource(getlist)}}'
     }
   }
}
```
![image](https://user-images.githubusercontent.com/24842116/118829118-6fa9e880-b8f0-11eb-9494-665600ddfca9.png)
![image](https://user-images.githubusercontent.com/24842116/118829170-77698d00-b8f0-11eb-9aa1-e5b3f0b8a762.png)

**Comment by liuweiGL** [comment]
你给个 demo 复现一下，不然我们没法帮助你。

**Comment by liuweiGL** [comment]
https://codesandbox.io/s/453qf  你可以使用这个作为模板来修改

**Comment by prgrmrwy** [comment]
找到原因了，上面的 SchemaRender 是通过 createSchemaField 得到的实例，通过实例传入 scope 可以在 x-reactions: "{{asd}}" 中正确调用，但是当在如下运行中使用的时候，只能拿到 createSchemaField 创建时传入的 scope
```javascript
{
  x-reactions: {
    dependencies: ['xxx'],
    fulfill: { run: '{{asd}}' }
  }
}
```
![image](https://user-images.githubusercontent.com/24842116/118831887-b6004700-b8f2-11eb-925e-1738eae7ad6c.png)

![image](https://user-images.githubusercontent.com/24842116/118831781-a254e080-b8f2-11eb-81fa-680c48059bae.png)

**Comment by liuweiGL** [comment, marked as answer]
可能你已经解决问题了，但还是建议你把有联动的字段封装成一个组件来使用：

```ts

import { useEffect } from 'react'

import useRequest from '@ahooksjs/use-request'
import { useForm } from '@formily/react'

type MockData = {
  a: number
  b: boolean
  c: string
}

const api = (params: any) => {}

const AsyncSelect = () => {
  const form = useForm<MockData>()

  const { a, b, c } = form.values

  const { data, run } = useRequest(api, {
    manual: true
  })

  useEffect(() => {
    run({
      a,
      b,
      c
    })
  }, [a, b, c, run])

  return (
    <Select>
      {data.map(item => (
        <Option />
      ))}
    </Select>
  )
}

export default observer(AsyncSelect)
```

**Reply by prgrmrwy** [reply]
https://codesandbox.io/s/exciting-field-l1wvl?file=/App.tsx

**Reply by prgrmrwy** [reply]
在 textarea 中，如果 run 里面使用 test 就会有引用错误，如果使用 asd 就是正常的，看起来是 run 的时候 scope 没有从实例里面获取

**Reply by liuweiGL** [reply]
@janryWang 你看看这个问题呗。

**Reply by liuweiGL** [reply]
这是个 bug，需要下个版本修复，你可以先用我上面的那种方式来实现。

**Reply by liuweiGL** [reply]
#1462

**Reply by prgrmrwy** [reply]
另一个问题：https://codesandbox.io/s/exciting-field-l1wvl?file=/App.tsx
即使 scope 提前处理好了(asd)，fulfill.run 仍然无法触发函数执行

**Reply by janryWang** [reply]
你都强制指定effect生命周期了，dependencies变化已经失效了的，没明白你到底要干啥，reactions原理是函数内部依赖了某个响应数据发生变化才会重复执行，麻烦文档看清楚点

**Reply by prgrmrwy** [reply]
1. 首先不知道你看不看得懂上面的代码，源码上写的很清楚，问题出在：run 里面 "{{asd}}" 的形式并不能执行，需要写成 "asd()" 才能正常执行（后面的写法我确实一开始没找到原因，后来查看源码发现了用法），不是生命周期或者依赖的问题
2. 既然你提到文档，这个是[官网说明](https://react.formilyjs.org/api/shared/schema#schemareactions)
```javascript
// effects 这里确实注明了主动模式下，但是没有说明优先级或者覆盖关系，理解上可以认为他可以
// 主动关联其他字段亦可被动关联其他字段，如果想要你说的这种能力，ts overload 了解一下？
effects?: SchemaReactionEffect[] //主动模式下的独立生命周期钩子
run?: string //执行语句 (你可以全文(上面官网链接)搜索下看看说明在什么位置)
```
```javascript
// 这是和 x-reactions 相关的两处示例
// 1. 
{
  "run": "$form.setFieldState('target',state=>{state.visible = $self.value === '123'})"
}
// 2.
{
    "x-reactions": "{{myReaction}}" //外部传入的函数，在函数内可以实现更复杂的联动
}
```
有没有看出区别？我是后来看到这里发现一个是执行，另一个只是引用，到后面看了下源码确认了这个奇怪的事：
run 的语义和 call 是类似的，你见过这样调用的? 
```javascript
function a(who) { console.log(`wow, ${who}`) };
call(a()) // ???
```
3. 说到底这是一个(不一定是)反模式的写法带来的理解误区，我只是在文档区没有分辨出差异的路人而已，这是一个 discuss 帖子，就是用来解惑的，不要求你文档精美完善，有问题请教就这态度？
4. 请不要一遇到问题就问甩文档，仔细看源码描述就知道是哪里的问题
> 如果想要实现一些更复杂的自定义组件，我们强烈推荐直接看@formily/antd或 @formily/next的源码

如上，我以为这么写的官网，自己看源码的能力也是一样的

**Reply by prgrmrwy** [reply]
> 这是个 bug，需要下个版本修复，你可以先用我上面的那种方式来实现。

1. 思路类似，可以解决
2. 也可以在 createSchemaField 的时候再传 scope

**Reply by janryWang** [reply]
注意，run的语义不是call，是执行一段语句，不是调用一个函数，个人觉得不属于反模式，如果真是反模式的话，可以晒一下是否有相关依据再判断，formily是接受喷设计问题的。

还有，如果你一开始的提问就像这样认真的提问，而不是提issue，issue要求也不看，提discussion连个代码用```都懒得用的态度，我也会认真思考你提的问题

> 1. 首先不知道你看不看得懂上面的代码，源码上写的很清楚，问题出在：run 里面 "{{asd}}" 的形式并不能执行，需要写成 "asd()" 才能正常执行（后面的写法我确实一开始没找到原因，后来查看源码发现了用法），不是生命周期或者依赖的问题
> 2. 既然你提到文档，这个是[官网说明](https://react.formilyjs.org/api/shared/schema#schemareactions)
> 
> ```js
> // effects 这里确实注明了主动模式下，但是没有说明优先级或者覆盖关系，理解上可以认为他可以
> // 主动关联其他字段亦可被动关联其他字段，如果想要你说的这种能力，ts overload 了解一下？
> effects?: SchemaReactionEffect[] //主动模式下的独立生命周期钩子
> run?: string //执行语句 (你可以全文(上面官网链接)搜索下看看说明在什么位置)
> ```
> 
> ```js
> // 这是和 x-reactions 相关的两处示例
> // 1. 
> {
>   "run": "$form.setFieldState('target',state=>{state.visible = $self.value === '123'})"
> }
> // 2.
> {
>     "x-reactions": "{{myReaction}}" //外部传入的函数，在函数内可以实现更复杂的联动
> }
> ```
> 
> 有没有看出区别？我是后来看到这里发现一个是执行，另一个只是引用，到后面看了下源码确认了这个奇怪的事：
> run 的语义和 call 是类似的，你见过这样调用的?
> 
> ```js
> function a(who) { console.log(`wow, ${who}`) };
> call(a()) // ???
> ```
> 
> 1. 说到底这是一个(不一定是)反模式的写法带来的理解误区，我只是在文档区没有分辨出差异的路人而已，这是一个 discuss 帖子，就是用来解惑的，不要求你文档精美完善，有问题请教就这态度？
> 2. 请不要一遇到问题就问甩文档，仔细看源码描述就知道是哪里的问题
> 
> > 如果想要实现一些更复杂的自定义组件，我们强烈推荐直接看@formily/antd或 @formily/next的源码
> 
> 如上，我以为这么写的官网，自己看源码的能力也是一样的

**Reply by prgrmrwy** [reply]
1. 我理解你的 issue 情绪，后面也转到 discuss 了，可以把情绪留到 bug 里面，就像上面给你转 bug 的人那样，后面的追问也是在 discuss 里面的，这就是个小的理解问题，你可以提更详细的要求(discuss)让我完善问题，不是直接甩文档
2. 我也说不一定是个反模式，英文语义本就有有些模糊，这里是指出使用时出现理解偏差导致提出上面问题
3. 上面已经截图给你看源码了，我相信诚意已经够了，况且如果你仔细看 [issue](https://github.com/alibaba/formily/issues/1459#issuecomment-844144315) 中你同事给的范例，也能看到我提的问题的原因
```javascript
{
   'x-reactions': {
      dependencies: ['xxx'],
      fulfill: {
          run: '{{useAsyncDataSource}}' // 这就是我问题的来源，这样不执行
     }
   }
}
```

**Reply by prgrmrwy** [reply]
又，discuss 没有 ``` 的问题，codepen 上已经写了，应该比 markdown 香一点吧

**Reply by janryWang** [reply]
开源社区文化是需要认真的态度的，如果没有这份认真只会变得一团糟，而且开源产品是需要社区一起共建的，即便是贡献修复文档的错别字，标点符号，都是非常值得称赞的，希望你能理解，如果你认可formily，完全欢迎加入进来共建，我也会对我之前的行为道歉

**Reply by prgrmrwy** [reply]
不用道歉，我很喜欢 formily，昨天撸源码也是想看看是不是可以贡献一把，没想到小哥操作迅速，只是上面回复感觉受到冒犯，实时对线一把，事说开了就完了

**Reply by janryWang** [reply]
你可以提个PR完善一下x-reactions effects的描述，你说的没说清楚覆盖的问题，这个的确有遗漏
